ddtrace/tracer/rules_sampler.go

```diff
@@ -0,0 +1,572 @@
+	// replacing '?' and '*' with regex characters
+	pattern = strings.Replace(pattern, "\\?", ".", -1)
+	pattern = strings.Replace(pattern, "\\*", ".*", -1)
+	//pattern must match an entire string
+	return regexp.Compile(fmt.Sprintf("^%s$", pattern))
+}
+
+// samplingRulesFromEnv parses sampling rules from the DD_TRACE_SAMPLING_RULES,
+// DD_SPAN_SAMPLING_RULES and DD_SPAN_SAMPLING_RULES_FILE environment variables.
+func samplingRulesFromEnv() ([]SamplingRule, error) {
+	var errs []string
+	var rules []SamplingRule
+	rulesFromEnv := os.Getenv("DD_TRACE_SAMPLING_RULES")
+	if rulesFromEnv != "" {
+		traceRules, err := unmarshalSamplingRules([]byte(rulesFromEnv), SamplingRuleTrace)
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+		rules = append(rules, traceRules...)
+	}
+	spanRules, err := unmarshalSamplingRules([]byte(os.Getenv("DD_SPAN_SAMPLING_RULES")), SamplingRuleSpan)
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+	rules = append(rules, spanRules...)
+	rulesFile := os.Getenv("DD_SPAN_SAMPLING_RULES_FILE")
+	if len(rules) != 0 {
+		if rulesFile != "" {
+			log.Warn("DIAGNOSTICS Error(s): DD_SPAN_SAMPLING_RULES is available and will take precedence over DD_SPAN_SAMPLING_RULES_FILE")
+		}
+		if len(errs) != 0 {
+			return rules, fmt.Errorf("%s", strings.Join(errs, "\n\t"))
+		}
+		return rules, nil
+	}
+	if rulesFile != "" {
+		rulesFromEnvFile, err := os.ReadFile(rulesFile)
+		if err != nil {
+			errs = append(errs, fmt.Sprintf("Couldn't read file from DD_SPAN_SAMPLING_RULES_FILE: %v", err))
+		}
+		spanRules, err = unmarshalSamplingRules(rulesFromEnvFile, SamplingRuleSpan)
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+		rules = append(rules, spanRules...)
+	}
+	if len(errs) != 0 {
+		return rules, fmt.Errorf("%s", strings.Join(errs, "\n\t"))
+	}
+	return rules, nil
+}
+
+// unmarshalSamplingRules unmarshals JSON from b and returns the sampling rules found, attributing
+// the type t to them. If any errors are occurred, they are returned.
+func unmarshalSamplingRules(b []byte, spanType SamplingRuleType) ([]SamplingRule, error) {
+	if len(b) == 0 {
+		return nil, nil
+	}
+	var jsonRules []struct {
+		Service      string      `json:"service"`
+		Name         string      `json:"name"`
+		Rate         json.Number `json:"sample_rate"`
+		MaxPerSecond float64     `json:"max_per_second"`
+	}
+	err := json.Unmarshal(b, &jsonRules)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshalling JSON: %v", err)
+	}
+	rules := make([]SamplingRule, 0, len(jsonRules))
+	var errs []string
+	for i, v := range jsonRules {
+		if v.Rate == "" {
+			errs = append(errs, fmt.Sprintf("at index %d: rate not provided", i))
+			continue
+		}
+		rate, err := v.Rate.Float64()
+		if err != nil {
+			errs = append(errs, fmt.Sprintf("at index %d: %v", i, err))
+			continue
+		}
+		if !(rate >= 0.0 && rate <= 1.0) {
+			log.Warn("at index %d: ignoring rule %+v: rate is out of [0.0, 1.0] range", i, v)
```

Comment:
why are we sometimes appending the error to `errs` and sometimes logging a warning, when they have the same side effect (i.e. continue the loop and ignore the rule)?